smtp-server/src/lib.rs

```diff
@@ -143,6 +154,45 @@ pub trait Config: Send + Sync {
     where
         R: Send + Unpin + AsyncRead;
 
+    /// `handle_mail_multi` is the alternative version of `handle_mail` that
+    /// must be implemented by LMTP servers: instead of being an async function
+    /// that returns a single decision, it is an async function that returns an
+    /// async stream of decisions that will all be sent back to the client.
+    /// Contrary to SMTP that returns a single decision signaling whether the
+    /// mail was accepted and added to the queue, in LMTP there must be one such
+    /// decision for each RCPT TO command that succeeded (i.e. for each accepted
+    /// recipient), which allows to indicate that the message could be stored in
+    /// the mailbox of some users, but not in that of other users.  This can
+    /// happen for instance if their mail quota is used up.
+    ///
+    /// This function is an async function that returns an async stream. The
+    /// lifetimes of the borrow on the mail's data stream is  limited to the
+    /// duration of the async call; this reference may not be retained by the
+    /// returned stream. The recommended implementation of this function would
+    /// start by reading the message's content in RAM (or to a temporary file),
+    /// and then produce a stream that writes the message to all of the
+    /// mailboxes one after the other.
+    ///
+    /// The default implementation of this function calls `handle_mail` to
+    /// produce a stream of only a single response. LMTP server implementors
+    /// must implement this function, and must define `handle_mail` as
+    /// `unreachable!()`.
```

Comment:
I'd feel a little better unifying the API between handle_mail_multi and handle_mail, having a single one (with the API of handle_mail_multi) and hinting at futures::stream::once in doc-comments about how to implement. Does that make sense to you?

The reason why I feel a bit uncomfortable with the way this is currently written is handle_mail and handle_mail_multi are a conscious choice that the crate user must make; whereas having a single one with a comment saying how many answers should be there depending on the protocol chosen means there's no hesitation to be had.

It'd be even better if we could statically enforce the right number of returns (or at least 1 return for smtp servers), but I don't think it's possible, so it's not a big deal IMO if smtp server writers just have to add one futures::stream::once call there.

Reply:
Your concern makes sense to me! But I'm thinking the clarification in terms of user experience (ie. "what should I do to implement Config?") beats the fact that there's less boilerplate in your first version :)

That said, just to check, it seems to me that you've currently implemented the thing with the Protocol being an associated constant. If it works it's the best of both worlds, but I also thought it'd require GATs?

Reply:
I actually found a way to do wighout GATs, although it isn't the prettiest ^^